### PR TITLE
[MLOps 1.5] Autogenerate data logging configs from the registry

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Feature Preview:
```python
string_result = data_logging_config_from_predefined(
        group_names="image_classification",
        frequency=2
        loggers={"some_logger": {"arg_1": 10}},
    )
```
outputs:
```bash
loggers:
  python: null

data_logging:
  pipeline_outputs.labels:
  - func: predicted_classes
    frequency: 1
  - func: predicted_top_score
    frequency: 1
  pipeline_inputs.images:
  - func: image_shape
    frequency: 1
  - func: mean_pixels_per_channel
    frequency: 1
  - func: std_pixels_per_channel
    frequency: 1
  - func: fraction_zeros
    frequency: 1
```

## Description
Introduces a new helper function, `data_logging_config_from_predefined`, that generates data logging configurations using metric group names and optional arguments. The function enables developers to quickly generate examples of data logging configurations, as demonstrated in the examples/data-logging-configs directory.

This PR includes comprehensive testing and features a test that, when run with the appropriate environment variable, can regenerate a fresh set of examples. This functionality is useful for maintaining up-to-date configurations as the log structure evolves over time. :
```bash
GENERATE_CONFIGS=1 pytest tests/deepsparse/loggers/metric_functions/helpers/test_data_config_from_predefined.py 
```

### CODE REVIEW

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+__all__ = ["data_logging_config_from_predefined"]
+
+from typing import Dict, List, Optional, Union, Any
+
+from deepsparse.loggers.build_logger import default_logger
+
+
+def data_logging_config_from_predefined(
+    group_names: Union[str, List[str]],
+    loggers: Dict[str, Optional[Dict[str, Any]]] = default_logger(),
```
Comment (bfineran) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:10]: been leaving this comment a lot - we cannot use mutable types as function defaults in python as they can have unintended side effects. read more here: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

**Review by bfineran — COMMENTED** [review]
@dbogunowicz great job splitting out the work on this feature
1. can you provide a sketch of how `data_logging_config_from_predefined` interacts with the registry?
2. this is still a 1.5 feature (see PR title)

**Comment by dbogunowicz** [comment]
@bfineran updated the code to demonstrate how the information is being fetched from the registry. The resulting `data_logging_config` can be now fed to the function `_nested_dict_to_lines` and this will return a full YAML configuration :)

File: src/deepsparse/loggers/metric_functions/helpers/config_generation.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+from deepsparse.loggers.build_logger import predefined_metric_function
+from deepsparse.loggers.config import MetricFunctionConfig
+from deepsparse.loggers.metric_functions.registry import DATA_LOGGING_REGISTRY
+
+
+_WHITESPACE = "  "
+_LOGGER = logging.getLogger(__name__)
+
+
+def data_logging_config_from_predefined(
```
Comment (corey-nm) [inline src/deepsparse/loggers/metric_functions/helpers/config_generation.py:37]: Was having pre-built yaml files that we edit by hand discussed as an alternative to this approach? It'd be a lot easier to edit a single yaml file than tweak these functions and have to double check that none of the other auto-generated yamls have changed. Especially if each of the yamls has slight differences in the structure that are hard to capture in this kind of serialization code

Comment (bfineran) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:37]: they were, ironically went with auto for the purpose of easier maintenance. And the hope was for the auto generated configs to be used as examples. 

@corey-nm (or @dbogunowicz ) do we expect these functions to have differences in structure in a way that the auto generated wouldn't capture? most functions should just be a name, target pipeline, and target identifier, no?

Do you see a difference between needing to manually edit manual files vs auto generating when an expected change occurs?

Additionally we can add tests to compare fresh auto generated configs vs the example docs that way GHA can catch and remind us to auto generate when they become stale or, if an unintended change was made that affects the config, a failed test would flag this as well (@dbogunowicz thoughts on this test)

Comment (dbogunowicz) [reply src/deepsparse/loggers/metric_functions/helpers/config_generation.py:37]: Let's sync on that offline.

**Review by bfineran — APPROVED** [review]

**Review by corey-nm — APPROVED** [review]
the yamls look good to me!

**Pull request merged**